File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/DynamicSerializer.java
package it.unibo.coordination.linda.presentation;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.Writer;
import java.util.Collection;

abstract class DynamicSerializer<T> extends SimpleSerializer<T> {


    public DynamicSerializer(MIMETypes mimeType, ObjectMapper mapper) {
        super(mimeType, mapper);
    }

    @Override
    public abstract Object toDynamicObject(T object);

    @Override
    public void write(T object, Writer writer) {
        writeImpl(toDynamicObject(object), writer);
    }

    @Override
    public void write(Collection<? extends T> objects, Writer writer) {
        writeImpl(toDynamicObject(objects), writer);
    }
}

File: presentation/src/main/kotlin/it/unibo/presentation/DynamicSerializer.kt
package it.unibo.presentation

import com.fasterxml.jackson.databind.ObjectMapper
import java.io.Writer

internal abstract class DynamicSerializer<T>(mimeType: MIMETypes, mapper: ObjectMapper)
    : SimpleSerializer<T>(mimeType, mapper) {

    abstract override fun toDynamicObject(`object`: T): Any

    override fun write(`object`: T, writer: Writer) {
        writeImpl(toDynamicObject(`object`), writer)
    }

    override fun write(objects: Collection<T>, writer: Writer) {
        writeImpl(toDynamicObject(objects), writer)
    }
}